src/main/java/org/lantern/event/PublicIpAndTokenTracker.java

```diff
@@ -10,58 +11,52 @@
 import com.google.inject.Singleton;
 
 /**
- * This class centralizes tracking of the combination of a valid refresh 
- * token and a valid proxy. Several services rely on that combination,
- * including the XMPP server login and connecting to the friends API.
+ * This class centralizes tracking of the combination of a valid refresh token
+ * and a valid proxy. Several services rely on that combination, including the
+ * XMPP server login and connecting to the friends API.
  */
 @Singleton
 public class PublicIpAndTokenTracker {
 
     private final Logger log = LoggerFactory.getLogger(getClass());
-    
-    private boolean hasRefresh = false;
 
-    private boolean hasPublicIp = false;
+    private volatile boolean hasPublicIp = false;
+    private volatile String refresh;
 
-    private String refresh;
-    
     @Inject
     public PublicIpAndTokenTracker(final Model model) {
         log.debug("Creating tracker");
         Events.register(this);
         this.refresh = model.getSettings().getRefreshToken();
-        if (StringUtils.isNotBlank(refresh)) {
-            this.hasRefresh = true;
-        }
-    }
-    
-    /**
-     * Resets the state of the tracker (useful when reinitializing system).
-     */
-    public void reset() {
-        this.hasPublicIp = false;
     }
     
     @Subscribe
-    public void onRefreshToken(final RefreshTokenEvent refreshEvent) {
-        log.debug("Got refresh token -- loading friends");
-        this.refresh = refreshEvent.getRefreshToken();
-        synchronized (this) {
-            this.hasRefresh = true;
-            if (this.hasPublicIp) {
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onPublicIp(final PublicIpEvent pce) {
+        log.debug("Got public IP");
+        this.hasPublicIp = true;
+        if (StringUtils.isNotBlank(refresh)) {
+            log.debug("Have both public IP and refresh token, posting event");
+            Events.asyncEventBus()
+                    .post(new PublicIpAndTokenEvent(this.refresh));
         }
     }
-    
+
     @Subscribe
-    public void onPublicIp(final PublicIpEvent pce) {
-        synchronized (this) {
-            this.hasPublicIp = true;
-            if (this.hasRefresh) {
-                log.debug("Posting event!!");
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onConnectivityChanged(ConnectivityChangedEvent cce) {
```

Comment:
So, PublicIpInfoHandler does the public ip check in response to a ProxyConnectionEvent, which happens as soon as the local proxy has started.

The reset call happens in startNetworkServices, which is called in response to a ConnectivityChangedEvent.  So I guess really the race here is between the connectivity check and the local proxy coming up.

In any case, making PublicIpAndTokenTracker only lose the stored IP in response to lost connectivity should give us what we need.  Previously, we would have stopped network services on the connectivity lost, and then cleared out the IP on connectivity being reestablished.  That has its own race between the connectivity event being delivered and the public ip event showing up.

I should note that if you look at the ticket, you'll see logging that clearly shows reset being called after the public IP has been captured, confirming that this is the root cause.  I should also note in my testing, with my change, I haven't been to reproduce the hang on the friends screen.

I can't test the scenario of connectivity being lost and reestablished because of #2198.